### PR TITLE
Add XFail support

### LINKED ISSUE #158: Add XFail support
Sometimes we have a testcase blocked by bug. To keep CI green we need to temporarily disable it with, for example, `skip_if()` function. However to avoid forever disabled testcases one should periodically check bug and remove `skip_if` when bug has been resolved as fixed. These can be avoided with xfail status. Behaviour of XFail similar to skip, but testcase with xfail should always run. If testcase has failed then report XFail status with a reason, if testcase has passed then report FAIL status and remind to remove xfail in a testcase code.

### PR DESCRIPTION
**Pull request by Steap2448**
Test can be marked as xfail. It will be treated as failed when succeeded.

I didn't forget about

- [x] Tests
- [x] Changelog
- [x] Documentation

Close #158

### CODE REVIEW

**Comment by ligurio** [comment]
AFAIU it is not possible to specify a reason of XFail and show that reason near the test status in output. Am I right?

**Comment by Steap2448** [comment]
> AFAIU it is not possible to specify a reason of XFail and show that reason near the test status in output. Am I right?

It is possible now.

**Comment by Steap2448** [comment]
> AFAIU it is not possible to specify a reason of XFail and show that reason near the test status in output. Am I right?

It is possible now.

File: luatest/assertions.lua [hunk 1]
```diff
@@ -15,6 +15,16 @@ local prettystr_pairs = pp.tostring_pair
...
+local mt = {}
+setmetatable(M, mt)
+
+mt.xfail = false
+function mt.is_xfail()
+    local xfail = mt.xfail
+    mt.xfail = false
+    return xfail
+end
+
```
Comment (rosik) [inline luatest/assertions.lua:27]: I don't like the idea of using metatables as private storage. Look, there's `M.private` definition two lines below. Let's put the `is_xfail` function there?

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+- Add `xfail` option
```
Comment (rosik) [inline CHANGELOG.md:5]:
```suggestion
- Add `xfail` test marks support.
```

File: README.rst [hunk 3]
```diff
@@ -219,6 +219,10 @@ List of luatest functions
 | ``fail_if (condition, message)``                                   | Stops a test due to a failure if condition    |
 |                                                                    | is met.                                       |
 +--------------------------------------------------------------------+-----------------------------------------------+
+| ``xfail (message)``                                                | Mark test as xfail.                           |
++--------------------------------------------------------------------+-----------------------------------------------+
+| ``xfail_if (condition, message)``                                  | Mark test as xfail if condition is met.       |
++--------------------------------------------------------------------+-----------------------------------------------+
```
Comment (rosik) [inline README.rst:225]:
```suggestion
| ``xfail (message)``                                                | Mark test as :ref:`XFail`.                    |
+--------------------------------------------------------------------+-----------------------------------------------+
| ``xfail_if (condition, message)``                                  | Mark test as :ref:`XFail` if condition is met.|
+--------------------------------------------------------------------+-----------------------------------------------+
```

File: README.rst [hunk 4]
```diff
@@ -237,6 +241,24 @@ List of luatest functions
...
+
+.. code-block:: Lua
+
+    local g = t.group()
+
+    g.test_fail = function()
+        t.xfail('Must fail no matter what')
+        t.assert_equals(3, 4)
+    end
+
```
Comment (rosik) [inline README.rst:261]:
```suggestion
.. _XFail

---------------------------------
XFail
---------------------------------

The ``xfail`` mark makes test results to be interpreted vice versa: it's
threated as passed when an assertion fails, and it fails if no errors are
raised. It allows one to mark a test as temporarily broken due to a bug in some
other component which can't be fixed immediately. It's also a good practice to
keep xfail tests in sync with an issue tracker.

.. code-block:: Lua

    local g = t.group()

    g.test_fail = function()
        t.xfail('Must fail no matter what')
        t.assert_equals(3, 4)
    end

XFail only applies to the errors raised by the luatest assertions. Regular Lua
errors still cause the test failure.

```

**Review by rosik — APPROVED** [review]

**Pull request merged**